Client/Frontend/Sync/BraveCore/Bookmarkv2.swift

```diff
@@ -0,0 +1,370 @@
+            
+            //Moving to the very last index.. same as appending..
+            if destinationIndexPath.row == parent.children.count - 1 {
+                node.move(toParent: parent)
+            } else {
+                node.move(toParent: parent, index: UInt(destinationIndexPath.row))
+            }
+        }
+    }
+}
+
+protocol BookmarksV2FetchResultsDelegate: class {
+    func controllerWillChangeContent(_ controller: BookmarksV2FetchResultsController)
+    
+    func controllerDidChangeContent(_ controller: BookmarksV2FetchResultsController)
+    
+    func controller(_ controller: BookmarksV2FetchResultsController, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?)
+    
+    func noIdeaReloadTable(_ controller: BookmarksV2FetchResultsController)
+}
+
+protocol BookmarksV2FetchResultsController {
+    /* weak */ var delegate: BookmarksV2FetchResultsDelegate? { get set }
+    
+    var fetchedObjects: [Bookmarkv2]? { get }
+    func performFetch() throws
+    func object(at indexPath: IndexPath) -> Bookmarkv2?
+}
+
+class Bookmarkv2Fetcher: NSObject, BookmarksV2FetchResultsController {
+    weak var delegate: BookmarksV2FetchResultsDelegate?
+    private var bookmarkModelListener: BookmarkModelListener?
+    private weak var bookmarksAPI: BraveBookmarksAPI?
+    
+    private let parentNode: BookmarkNode?
+    private var children = [BookmarkNode]()
+    
+    init(_ parentNode: BookmarkNode?, api: BraveBookmarksAPI) {
+        self.parentNode = parentNode
+        self.bookmarksAPI = api
+        super.init()
+        
+        self.bookmarkModelListener = api.add(BookmarkModelStateChangeObserver { [weak self] _ in
+            guard let self = self else { return }
+            self.delegate?.noIdeaReloadTable(self)
+        })
+    }
+    
+    var fetchedObjects: [Bookmarkv2]? {
+        return children.map({ Bookmarkv2($0) })
+    }
+    
+    func performFetch() throws {
+        self.children.removeAll()
+        
+        if let parentNode = self.parentNode {
+            self.children.append(contentsOf: parentNode.children)
+        } else {
+            if let node = bookmarksAPI?.mobileNode {
+                self.children.append(node)
+            }
+            
+            if let node = bookmarksAPI?.desktopNode, !node.children.isEmpty {
+                self.children.append(node)
+            }
+            
+            if let node = bookmarksAPI?.otherNode, !node.children.isEmpty {
+                self.children.append(node)
+            }
+            
+            if self.children.isEmpty {
+                throw NSError(domain: "brave.core.migrator", code: -1, userInfo: [
+                    NSLocalizedFailureReasonErrorKey: "Invalid Bookmark Nodes"
+                ])
+            }
+        }
+    }
+    
+    func object(at indexPath: IndexPath) -> Bookmarkv2? {
+        return Bookmarkv2(children[indexPath.row])
```

Comment:
Add this for index out of bounds safety
`guard let children = children[safe: indexPath.row] else { return nil }`